File: src/installer/elevator.ts
import fs from 'fs';

import { Logger } from './logger';
import { lsControl } from './luna';

const logger = new Logger('elevator');

const CLIENT_PERMISSIONS_PATH = `/var/luna-service2-dev/client-permissions.d/${process.env.APP_ID}.all.json`;
const MANIFEST_PATH = `/var/luna-service2-dev/manifests.d/${process.env.APP_ID}.json`;

const addExtraClientPermissions = () => {
	if (fs.existsSync(CLIENT_PERMISSIONS_PATH)) {
		logger.info('client permissions file already exists');
		return;
	}

	fs.writeFileSync(CLIENT_PERMISSIONS_PATH, JSON.stringify({
		[`${process.env.APP_ID}-*`]: [
			'all',
		],
	}));

	logger.info(`created client permissions file: ${CLIENT_PERMISSIONS_PATH}`);
};

const updateManifest = () => {
	const manifest = JSON.parse(fs.readFileSync(MANIFEST_PATH).toString('utf8'));

	manifest.clientPermissionFiles ??= [];
	manifest.clientPermissionFiles.push(CLIENT_PERMISSIONS_PATH);

	fs.writeFileSync(MANIFEST_PATH, JSON.stringify(manifest));

	logger.info(`updated manifest file: ${MANIFEST_PATH}`);
};

export const elevate = () => {
	addExtraClientPermissions();
	updateManifest();

	lsControl('scan-volatile-dirs');
	lsControl('scan-services');
};

File: agent/src/routines/elevator.rs
use std::{fs, io};
use std::collections::HashMap;
use std::path::Path;

use serde_json::{json, Value};
use serde::{Serialize, Deserialize};

use crate::luna;
use crate::luna::LunaHubControlAction;

#[derive(Serialize, Deserialize, Debug)]
struct Manifest {
    #[serde(rename = "clientPermissionFiles")]
    client_permission_files: Vec<String>,

    #[serde(flatten)]
    rest: HashMap<String, Value>,
}

fn rescan() {
    luna::control(LunaHubControlAction::ScanVolatileDirs);
    luna::control(LunaHubControlAction::ScanServices);
}

pub fn elevate(app_id: &str) -> Result<(), io::Error> {
    let client_permissions_path =
        format!("/var/luna-service2-dev/client-permissions.d/{}.all.json", app_id);
    let manifest_path =
        format!("/var/luna-service2-dev/manifests.d/{}.json", app_id);

    if !Path::new(client_permissions_path.as_str()).exists() {
        let client_permissions = json!({
            format!("{}-*", app_id): [
                "all",
            ],
        });

        fs::write(&client_permissions_path, client_permissions.to_string())?;

        println!("Created extra client permissions file.");
    }

    let manifest = fs::read_to_string(&manifest_path)?;
    let mut manifest: Manifest = serde_json::from_str(manifest.as_str())?;

    if manifest.client_permission_files.contains(&client_permissions_path) {
        return Ok(());
    }

    manifest.client_permission_files.push(client_permissions_path);
    manifest.client_permission_files.sort_unstable();
    manifest.client_permission_files.dedup();

    let manifest = serde_json::to_string(&manifest)?;

    fs::write(&manifest_path, manifest)?;

    println!("Rewired manifest file.");

    rescan();

    Ok(())
}
